packages/@aws-cdk/cdk/lib/physical-name.ts

```diff
@@ -0,0 +1,130 @@
+}
+
+/**
+ * Represents the physical (that is, visible in the AWS Console) name of a resource.
+ * These can be known at code runtime,
+ * in the case of fixed, customer-provided names,
+ * at synthesis time, in case of automatically assigned names by the framework,
+ * or only at deploy time, assigned by CloudFormation -
+ * which is the default.
+ *
+ * @see #auto()
+ * @see #of()
+ */
+export abstract class PhysicalName {
+  /**
+   * A physical name that will be automatically generated.
+   * The details of how and when that generation will occur can be customized by passing an instance
+   * of {@link AutoPhysicalNameOptions}.
+   * By default, the physical name of the resource will be generated by CloudFormation,
+   * at deploy time.
+   */
+  public static auto(options: AutoPhysicalNameOptions = {}): PhysicalName {
+    return new LateBoundPhysicalName(options);
+  }
+
+  /**
+   * A fixed physical name (one that is known statically, at synthesis time).
+   *
+   * @param name the name to assign
+   */
+  public static of(name: string): PhysicalName {
+    return new FixedPhysicalName(name);
+  }
+
+  protected constructor() {
+  }
+
+  /**
+   * Returns the physical name, as a String.
+   * If the customer assigned a physical name to the resource, it will be returned, unchanged;
+   * can be `undefined`, if a physical name was not provided,
+   * or an opaque Token, if a generated name should be assigned lazily at synthesis time.
+   */
+  public abstract get value(): string | undefined;
+
+  /**
+   * A callback method called if the resource this physical name
+   * belongs to has been used in a cross-environment fashion.
+   * Allows late-binding of the actual name to only be generated if necessary,
+   * and kept deploy-time if not.
+   *
+   * The default implementation does nothing,
+   * but can be overridden in subclasses.
+   *
+   * @param _resource the resource this physical name belongs to
+   * @internal
+   */
+  public _resolveCrossEnvironment(_resource: IResource): void {
+    // does nothing
+  }
+}
+
+// implementations are private to this module,
+// we only surface them through static factory methods
+
+class FixedPhysicalName extends PhysicalName {
+  private readonly name: string;
+
+  constructor(name: string) {
+    super();
+
+    this.name = name;
+  }
+
+  public get value(): string | undefined {
+    return this.name;
+  }
+}
+
+class LateBoundPhysicalName extends PhysicalName {
+  /**
+   * The value of the physical name -
+   * either a Token string if crossEnvironment is true,
+   * or undefined otherwise.
+   */
+  public readonly value: string | undefined;
+  private name?: string;
+
+  constructor(options: AutoPhysicalNameOptions) {
+    super();
+
+    this.value = options.crossEnvironment
+      ? new Token(() => this.name).toString()
+      : undefined;
+  }
+
+  /** @internal */
+  public _resolveCrossEnvironment(resource: IResource): void {
+    if (this.value) {
```

Comment:
For readability, bail out early:

    if (!value) {
      throw....
    }

    // success path

Reply:
Done.